include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -371,19 +419,136 @@ class pool_memory_resource final : public device_memory_resource {
+      ptds_events.push_back(e);
+      return e.event;
+    }
+#else
+    // We use cudaStreamLegacy as the event map key for the default stream for consistency between
+    // PTDS and non-PTDS mode. In PTDS mode, the cudaStreamLegacy map key will only exist if the
+    // user explicitly passes it, so it is used as the default location for the free list
+    // at construction, and for merging free lists when a thread exits (see destroy_event()).
+    // For consistency, the same key is used for null stream free lists in non-PTDS mode.
+    if (cudaStreamDefault == stream) { stream = cudaStreamLegacy; }
+#endif
+
+    auto iter = stream_events_.find(stream);
+    if (iter == stream_events_.end()) {
+      cudaEvent_t event{};
+      auto result = cudaEventCreateWithFlags(&event, cudaEventDisableTiming);
+      assert(cudaSuccess == result);
+      stream_events_[stream] = event;
+      event_streams_[event]  = stream;
+      return event;
+    } else {
+      return iter->second;
+    }
+  }
+
+  /**
+   * @brief Destroy the specified CUDA event and move all free blocks for the associated stream
+   * to the default stream free list.
+   *
+   * @param event The event to destroy.
+   */
+  void destroy_event(cudaEvent_t event)
+  {
+    // If we are destroying an event with associated free list, we need to synchronize that event
+    // and then merge its free list into the (legacy) default stream's list
+    auto free_list_iter = stream_free_blocks_.find(event);
+    if (free_list_iter != stream_free_blocks_.end()) {
+      auto blocks = free_list_iter->second;
+      stream_free_blocks_[get_event(cudaStreamLegacy)].insert(blocks.begin(), blocks.end());
+      stream_free_blocks_.erase(free_list_iter);
+
+      auto result = cudaEventSynchronize(event);
+      assert(cudaSuccess == result);
+    }
+    auto result = cudaEventDestroy(event);
+    assert(cudaSuccess == result);
+  }
+
   size_t maximum_pool_size_;
   size_t current_pool_size_{0};
 
   Upstream* upstream_mr_;  // The "heap" to allocate the pool from
 
   // map of [stream_id, free_list] pairs
   // stream stream_id must be synced before allocating from this list to a different stream
-  std::map<cudaStream_t, free_list> stream_free_blocks_;
+  // std::map<cudaStream_t, free_list> stream_free_blocks_;
+  std::map<cudaEvent_t, free_list> stream_free_blocks_;
 
   std::set<block, rmm::mr::detail::compare_blocks<block>> allocated_blocks_;
 
   // blocks allocated from upstream: so they can be easily freed
   std::vector<block> upstream_blocks_;
+
+  // bidirectional mapping between non-default streams and events
+  std::unordered_map<cudaStream_t, cudaEvent_t> stream_events_;
+  std::unordered_map<cudaEvent_t, cudaStream_t> event_streams_;
+
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+  // references to per-thread events to avoid use-after-free when threads exit after MR is deleted
+  std::list<std::reference_wrapper<cuda_event>> ptds_events;
```

Comment:
```suggestion
  std::list<std::reference_wrapper<cuda_event>> ptds_events_;
```
Should this have a trailing underscore? Same for `mtx`